Review followup: PR #18699 — GH#18681: decompose dispatch_enrichment_workers into focused helper functions

## Unaddressed review bot suggestions

PR #18699 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/18699

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `prompts/build.txt`
section 6a). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `prompts/build.txt` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/pulse-quality-debt.sh:197`
- `.agents/scripts/pulse-quality-debt.sh:333`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/pulse-quality-debt.sh:197`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/18699#discussion_r3075143900)

```diff
+# Exit codes:
+#   0 - data fetched (partial data is acceptable; fields may be empty strings)
+#######################################
+_enrichment_fetch_issue_data() {
+	local issue_number="$1"
+	local repo_slug="$2"
+
+	local issue_body issue_title issue_comments
+	issue_body=$(gh issue view "$issue_number" --repo "$repo_slug" \
+		--json body --jq '.body // ""' 2>/dev/null) || issue_body=""
+	issue_title=$(gh issue view "$issue_number" --repo "$repo_slug" \
+		--json title --jq '.title // ""' 2>/dev/null) || issue_title=""
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> For efficiency, the two `gh issue view` calls to fetch the issue title and body can be combined into a single API call. Additionally, consolidate the `jq` processing into a single pass to improve performance and script efficiency.
> 
> ```suggestion
> 	local issue_json issue_title issue_body
> 	issue_json=$(gh issue view "$issue_number" --repo "$repo_slug" --json title,body 2>/dev/null) || issue_json="{}"
> 	# Consolidate jq calls into a single pass using @sh for robust shell variable assignment
> 	eval "$(echo "$issue_json" | jq -r '@sh "issue_title=\(.title // \"\"); issue_body=\(.body // \"\")"')"
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Consolidate multiple 'jq' calls into a single pass where possible to improve performance and script efficiency.
> </details>

#### gemini-code-assist on `.agents/scripts/pulse-quality-debt.sh:333`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/18699#discussion_r3075143914)

```diff
+	local post_body
+	post_body=$(gh issue view "$issue_number" --repo "$repo_slug" \
+		--json body --jq '.body // ""' 2>/dev/null) || post_body=""
+
+	if [[ "$post_body" == *"Worker Guidance"* ]]; then
+		echo "[pulse-wrapper] Enrichment: successfully added Worker Guidance to #${issue_number} in ${repo_slug}" >>"$LOGFILE"
+		return 0
+	else
+		echo "[pulse-wrapper] Enrichment: worker ran (exit=${enrichment_exit}) but no Worker Guidance found in #${issue_number} body (${#post_body} chars)" >>"$LOGFILE"
+		return 1
+	fi
+}
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> To ensure robust resource cleanup, use the project's established pattern with `_save_cleanup_scope`, `trap '_run_cleanups' RETURN`, and `push_cleanup`. Additionally, temporary files should be removed as soon as they are no longer needed, and all functions must include explicit return statements.
> 
> ```shell
> _enrichment_run_worker() {
> 	local prompt_file="$1"
> 	local issue_number="$2"
> 	local repo_slug="$3"
> 	local repo_path="$4"
> 	local resolved_model="$5"
> 
> 	_save_cleanup_scope
> 	trap '_run_cleanups' RETURN
> 
> 	local enrichment_output
> 	enrichment_output=$(mktemp)
> 	push_cleanup "rm -f \"$enrichment_output\""
> 	push_cleanup "rm -f \"$prompt_file\""
> 
> 	# shellcheck disable=SC2086
> 	"$HEADLESS_RUNTIME_HELPER" run \
> 		--role worker \
> 		--session-key "enrichment-${issue_number}" \
> 		--dir "$repo_path" \
> 		--model "$resolved_model" \
> 		--title "Enrichment analysis: Issue #${issue_number}" \
> 		--prompt-file "$prompt_file" </dev/null >"$enrichment_output" 2>&1
> 
> 	local enrichment_exit=$?
> 
> 	# Remove temporary file as soon as it is no longer needed
> 	rm -f "$enrichment_output"
> 
> 	# Check if enrichment succeeded (issue body was edited)
> 	local post_body
> 	post_body=$(gh issue view "$issue_number" --repo "$repo_slug" \
> 		--json body --jq '.body // ""' 2>/dev/null) || post_body=""
> 
> 	if [[ "$post_body" == *"Worker Guidance"* ]]; then
> 		echo "[pulse-wrapper] Enrichment: successfully added Worker Guidance to #${issue_number} in ${repo_slug}" >>"$LOGFILE"
> 		rm -f "$prompt_file"
> 		return 0
> 	else
> 		echo "[pulse-wrapper] Enrichment: worker ran (exit=${enrichment_exit}) but no Worker Guidance found in #${issue_number} body (${#post_body} chars)" >>"$LOGFILE"
> 		rm -f "$prompt_file"
> 		return 1
> 	fi
> }
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. For resource cleanup in shell scripts, use the established project pattern: use _save_cleanup_scope, trap '_run_cleanups' RETURN, and push_cleanup for robust cleanup on any exit path, and also

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/18699#pullrequestreview-4101180495)

> ## Code Review
> 
> This pull request refactors the enrichment worker dispatch logic in `pulse-quality-debt.sh` by decomposing the `dispatch_enrichment_workers` function into several modular helper functions. Feedback suggests optimizing GitHub CLI usage by combining issue data fetches into a single call and adopting the project's standard resource cleanup pattern for temporary files.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.8.22 automated scan.

--- a/.agents/scripts/pulse-quality-debt.sh
+++ b/.agents/scripts/pulse-quality-debt.sh
@@ -190,11 +190,11 @@ _enrichment_fetch_issue_data() {
 	local issue_number="$1"
 	local repo_slug="$2"
 
-	local issue_body issue_title issue_comments
-	issue_body=$(gh issue view "$issue_number" --repo "$repo_slug" \
-		--json body --jq '.body // ""' 2>/dev/null) || issue_body=""
-	issue_title=$(gh issue view "$issue_number" --repo "$repo_slug" \
-		--json title --jq '.title // ""' 2>/dev/null) || issue_title=""
+	local issue_json issue_body issue_title issue_comments
+	issue_json=$(gh issue view "$issue_number" --repo "$repo_slug" \
+		--json title,body 2>/dev/null) || issue_json="{}"
+	issue_title=$(printf '%s' "$issue_json" | jq -r '.title // ""')
+	issue_body=$(printf '%s' "$issue_json" | jq -r '.body // ""')
 
 	# Get kill/dispatch comments for failure context
 	issue_comments=$(gh api "repos/${repo_slug}/issues/${issue_number}/comments" \
@@ -303,8 +303,13 @@ _enrichment_run_worker() {
 	local repo_path="$4"
 	local resolved_model="$5"
 
+	_save_cleanup_scope
+	trap '_run_cleanups' RETURN
+
 	local enrichment_output
 	enrichment_output=$(mktemp)
+	push_cleanup "rm -f '${enrichment_output}'"
+	push_cleanup "rm -f '${prompt_file}'"
 
 	# shellcheck disable=SC2086
 	"$HEADLESS_RUNTIME_HELPER" run \
@@ -316,7 +321,6 @@ _enrichment_run_worker() {
 		--prompt-file "$prompt_file" </dev/null >"$enrichment_output" 2>&1
 
 	local enrichment_exit=$?
-	rm -f "$prompt_file" "$enrichment_output"
 
 	# Check if enrichment succeeded (issue body was edited)
 	local post_body

